File: Prefix Sum/Longest Subarray with Sum K.py
class Solution:
    def longestSubarray(self, arr, k):  
        # code here
        mp={}
        res=0
        preSum=0
        
        for i in range(len(arr)):
            preSum+=arr[i]
            if preSum==k:
                res=i+1
            elif (preSum-k) in mp:
                res=max(res,i-mp[preSum-k])
            if preSum not in mp:
                mp[preSum]=i
        return res
    

File: Prefix Sum/C++/Longest Subarray with Sum K.cpp
class Solution {
  public:
    int longestSubarray(vector<int>& arr, int k) {
        // code here
        unordered_map<int, int> mp;
        int res = 0 ;
        int preSum = 0;
        for (int i = 0; i < arr.size(); i++) {
            preSum += arr[i];
            
            if (preSum == k)
                res = i + 1;
            else if (mp.find(preSum - k) != mp.end())
                res = max(res, i - mp[preSum - k]);
            
            if (mp.find(preSum) == mp.end())
                mp[preSum] = i;
        }
        return res;
    }
};
